pkg/apis/batch/validation/validation.go

```diff
@@ -179,6 +211,114 @@ func validateJobSpec(spec *batch.JobSpec, fldPath *field.Path, opts apivalidatio
 		allErrs = append(allErrs, field.NotSupported(fldPath.Child("template", "spec", "restartPolicy"),
 			spec.Template.Spec.RestartPolicy, []string{string(api.RestartPolicyOnFailure), string(api.RestartPolicyNever)}))
+	} else if spec.PodFailurePolicy != nil && spec.Template.Spec.RestartPolicy != api.RestartPolicyNever {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("template", "spec", "restartPolicy"),
+			spec.Template.Spec.RestartPolicy, fmt.Sprintf("only %q is supported when podFailurePolicy is specified", api.RestartPolicyNever)))
+	}
+	return allErrs
+}
+
+func validatePodFailurePolicy(spec *batch.JobSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	rulesPath := fldPath.Child("rules")
+	if len(spec.PodFailurePolicy.Rules) > maxPodFailurePolicyRules {
+		allErrs = append(allErrs, field.TooMany(rulesPath, len(spec.PodFailurePolicy.Rules), maxPodFailurePolicyRules))
+	} else {
+		containerNames := sets.NewString()
+		for _, containerSpec := range spec.Template.Spec.Containers {
+			containerNames.Insert(containerSpec.Name)
+		}
+		for _, containerSpec := range spec.Template.Spec.InitContainers {
+			containerNames.Insert(containerSpec.Name)
+		}
+		for i, rule := range spec.PodFailurePolicy.Rules {
+			allErrs = append(allErrs, validatePodFailurePolicyRule(&rule, rulesPath.Index(i), containerNames)...)
+		}
+	}
+	return allErrs
+}
+
+func validatePodFailurePolicyRule(rule *batch.PodFailurePolicyRule, rulePath *field.Path, containerNames sets.String) field.ErrorList {
+	allErrs := field.ErrorList{}
+	actionPath := rulePath.Child("action")
+	if rule.Action == "" {
+		allErrs = append(allErrs, field.Required(actionPath, fmt.Sprintf("valid values: %q", supportedPodFailurePolicyActions.List())))
+	} else if !supportedPodFailurePolicyActions.Has(string(rule.Action)) {
+		allErrs = append(allErrs, field.NotSupported(actionPath, rule.Action, supportedPodFailurePolicyActions.List()))
+	}
+	if rule.OnExitCodes != nil {
+		allErrs = append(allErrs, validatePodFailurePolicyRuleOnExitCodes(rule.OnExitCodes, rulePath.Child("onExitCodes"), containerNames)...)
+	}
+	if rule.OnPodConditions != nil {
+		allErrs = append(allErrs, validatePodFailurePolicyRuleOnPodConditions(rule.OnPodConditions, rulePath.Child("onPodConditions"))...)
+	}
+	if rule.OnExitCodes != nil && rule.OnPodConditions != nil {
+		allErrs = append(allErrs, field.Invalid(rulePath, field.OmitValueType{}, "specifying both OnExitCodes and OnPodConditions is not supported"))
+	}
+	if rule.OnExitCodes == nil && rule.OnPodConditions == nil {
+		allErrs = append(allErrs, field.Invalid(rulePath, field.OmitValueType{}, "specifying one of OnExitCodes and OnPodConditions is required"))
+	}
+	return allErrs
+}
+
+func validatePodFailurePolicyRuleOnPodConditions(onPodConditions []batch.PodFailurePolicyOnPodConditionsPattern, onPodConditionsPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
```

Comment:
`var allErrs field.ErrorList`

This way, you save an allocation if there are actually no errors.

Reply:
updated